File: Medium/1647. Minimum Deletions to Make Character Frequencies Unique/1647. Minimum Deletions to Make Character Frequencies Unique.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class Solution {
    public int minDeletions(String s) {
        Map<Character, Integer> freqMap = new HashMap<>();

        // Count the frequency of each character
        for (char c : s.toCharArray()) {
            freqMap.put(c, freqMap.getOrDefault(c, 0) + 1);
        }

        int deletions = 0;
        Set<Integer> usedFrequencies = new HashSet<>();

        for (int count : freqMap.values()) {
            int currentCount = count;
            while (usedFrequencies.contains(currentCount) && currentCount > 0) {
                currentCount--;
                deletions++;
            }

            if (currentCount > 0) {
                usedFrequencies.add(currentCount);
            }
        }

        return deletions;
    }
}

File: Medium/1647. Minimum Deletions to Make Character Frequencies Unique/1647. Minimum Deletions to Make Character Frequencies Unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        freq_map = {}

        # Count the frequency of each character
        for char in s:
            freq_map[char] = freq_map.get(char, 0) + 1

        deletions = 0
        used_frequencies = set()

        for count in freq_map.values():
            current_count = count
            while current_count in used_frequencies and current_count > 0:
                current_count -= 1
                deletions += 1

            if current_count > 0:
                used_frequencies.add(current_count)

        return deletions
